### PR TITLE
Fix Service.OwnedResources() to return correct result

### LINKED ISSUE #966: `detectBindingResources` is binding to a wrong Service
### What is the environment (Minikube, Openshift)?
minikube

### What is the SBO version used?
v0.7.1

#### 1. deployed sample deployment and operator and operand into the same namespace called `test`
  
<details>
 <summary>deployment.yaml</summary>

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: busybox
  namespace: test
spec:
  replicas: 1
  selector:
    matchLabels:
      app: busybox
  template:
    metadata:
      labels:
        app: busybox
    spec:
      containers:
      - command:
        - sleep
        - infinity
        image: busybox
        name: busybox
```
</details>

<details>
 <summary>operator.yaml</summary>

```yaml

apiVersion: operators.coreos.com/v1
kind: OperatorGroup
metadata:
  name: operatorgroup
  namespace: test
spec:
  targetNamespaces:
  - test
---
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: my-postgresql-operator-dev4devs-com
  namespace: test
spec:
  channel: alpha
  name: postgresql-operator-dev4devs-com
  source: operatorhubio-catalog
  sourceNamespace: olm
```
</details>

<details>
 <summary>operand.yaml</summary>

```yaml
apiVersion: postgresql.dev4devs.com/v1alpha1
kind: Database
metadata:
  name: postgresql
  namespace: test
spec:
  databaseName: example
  databasePassword: postgres
  databaseStorageRequest: 1Gi
  databaseUser: postgres
  image: centos/postgresql-96-centos7
  size: 1

```
</details>

#### 2. created SericeBinding with `detectBindingResources: true`
<details>
 <summary>binding.yaml</summary>

```yaml
apiVersion: binding.operators.coreos.com/v1alpha1
kind: ServiceBinding
metadata:
  name: busybox-postgresql
  namespace: test
spec:
  application:
    group: apps
    name: busybox
    resource: deployments
    version: v1
  bindAsFiles: false
  detectBindingResources: true
  services:
    - group: postgresql.dev4devs.com
      kind: Database
      name: postgresql
      namespace: test
      version: v1alpha1
```
</details>

In my namespace, there are two Services

```
▶ k get svc
NAME                          TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)             AGE
postgresql                    ClusterIP   10.110.53.111   <none>        5432/TCP            14m
postgresql-operator-metrics   ClusterIP   10.99.217.236   <none>        8383/TCP,8686/TCP   14m
```

`postgresql-operator-metrics` Service  has operator (`Database.postgresql.dev4devs.com`) as `ownerReference`

```
▶ k get svc postgresql-operator-metrics -o jsonpath="{.metadata.ownerReferences}" | jq
[
  {
    "apiVersion": "apps/v1",
    "blockOwnerDeletion": true,
    "controller": true,
    "kind": "Deployment",
    "name": "postgresql-operator",
    "uid": "8c11676c-fb8d-4881-8f71-ddcba37581f2"
  }
]
```

`postgresql` Service has operand (`Database.postgresql.dev4devs.com`) as `ownerReference`
```
▶ k get svc postgresql -o jsonpath="{.metadata.ownerReferences}" | jq                 
[
  {
    "apiVersion": "postgresql.dev4devs.com/v1alpha1",
    "blockOwnerDeletion": true,
    "controller": true,
    "kind": "Database",
    "name": "postgresql",
    "uid": "43149969-d701-460a-8566-4337cbe4cc50"
  }
]
```

### What is the expected behaviour?
Deployment `busybox` has `DATABASE_CLUSTERIP` env variable containing IP address of the **`postgresql`** Service

### What is the actual behaviour?

Deployment `busybox` has `DATABASE_CLUSTERIP` env variable containing IP address of the **`postgresql-operator-metrics`** Service which is not tied to the `service` defined in `ServiceBinding` 

```
▶ k exec -it busybox-75fcb776bc-5m4hr -- env | grep CLUSTERIP                                         
DATABASE_CLUSTERIP=10.99.217.236
```

```
▶ k get svc
NAME                          TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)             AGE
postgresql                    ClusterIP   10.110.53.111   <none>        5432/TCP            14m
postgresql-operator-metrics   ClusterIP   10.99.217.236   <none>        8383/TCP,8686/TCP   14m
```

### Service Binding Operator Logs

<details>
 <summary>logs</summary>

```
{"level":"info","ts":1621435172.1368315,"logger":"controller-runtime.metrics","msg":"metrics server is starting to listen","addr":":8080"}
{"level":"info","ts":1621435172.1373563,"logger":"setup","msg":"starting manager"}
{"level":"info","ts":1621435172.1376226,"logger":"controller-runtime.manager","msg":"starting metrics server","path":"/metrics"}
I0519 14:39:32.137605       1 leaderelection.go:243] attempting to acquire leader lease  operators/8fa65150.coreos.com...
I0519 14:39:32.143465       1 leaderelection.go:253] successfully acquired lease operators/8fa65150.coreos.com
{"level":"info","ts":1621435172.1436667,"logger":"controller","msg":"Starting EventSource","reconcilerGroup":"binding.operators.coreos.com","reconcilerKind":"ServiceBinding","controller":"servicebinding","source":"kind source: /, Kind="}
{"level":"info","ts":1621435172.2441118,"logger":"controller","msg":"Starting Controller","reconcilerGroup":"binding.operators.coreos.com","reconcilerKind":"ServiceBinding","controller":"servicebinding"}
{"level":"info","ts":1621435172.244153,"logger":"controller","msg":"Starting workers","reconcilerGroup":"binding.operators.coreos.com","reconcilerKind":"ServiceBinding","controller":"servicebinding","worker count":1}
{"level":"info","ts":1621436719.492417,"logger":"controllers.ServiceBinding","msg":"Reconciling","serviceBinding":"test/busybox-postgresql","sb":{"apiVersion":"binding.operators.coreos.com/v1alpha1","kind":"ServiceBinding","namespace":"test","name":"busybox-postgresql"}}
{"level":"info","ts":1621436720.3018339,"logger":"controllers.ServiceBinding","msg":"Done","serviceBinding":"test/busybox-postgresql","retry":false,"error":null}

```
</details>

### PR DESCRIPTION
**Pull request by pedjak**
We were bitten by https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable

Fixed the loop to use the new variable.

Fixes #966

### CODE REVIEW

**Comment by baijum** [comment]
If no plan to create an acceptance test as part of this PR, please create an issue to track it.

File: pkg/reconcile/pipeline/context/service_test.go [hunk 1]
```diff
@@ -43,8 +47,55 @@ var _ = Describe("Service", func() {
 		Expect(res).To(BeNil())
 	})
 
+	It("should return owned resources", func() {
```
Comment (baijum) [inline pkg/reconcile/pipeline/context/service_test.go:50]: There is a scenario where the `OwnedResources` could return an error. A unit test would be nice.
Use this function to return error: https://pkg.go.dev/k8s.io/client-go/testing#Fake.PrependReactor
There is an example in PR #971

File: pkg/reconcile/pipeline/context/service_test.go [hunk 2]
```diff
@@ -43,8 +47,55 @@ var _ = Describe("Service", func() {
...
+		u.SetName("foo")
+		u.SetNamespace(ns)
+		u.SetUID(id)
+
+		objs = append(objs, u)
+
+		var children []interface{}
+
+		for _, gvr := range bindableResourceGVRs {
+			gvk := gvr.GroupVersion().WithKind(strings.Title(gvr.Resource)[:len(gvr.Resource)-1])
```
Comment (baijum) [inline pkg/reconcile/pipeline/context/service_test.go:66]: The title form `Configmap` is not equivalent to `ConfigMap`. But I think it's ok for testing. An acceptance test would be better to capture a real-world scenario.

Comment (pedjak) [reply pkg/reconcile/pipeline/context/service_test.go]: good catch - I have modified the test to exclude configmaps and check additional path when `List` returns not found error.

File: pkg/reconcile/pipeline/context/service_test.go (hunk 1 above)
Comment (pedjak) [reply pkg/reconcile/pipeline/context/service_test.go]: Added.

**Comment by pedjak** [comment]
> If no plan to write an acceptance test as part of this PR, please create an issue to track it.


@baijum  IMHO, constructing a reliable, deterministic acceptance test for this bug is impossible, because the bug is triggered by the order of resources returned by cluster API, and we cannot control that in practice. Fortunately we can control the order in the unit tests.

I could agree that writing additional acceptance tests demonstrating `detectBindingResource` usage would beneficial, but IMHO it is out of the scope of this PR.

**Comment by baijum** [comment]
/lgtm

Not adding "approve" now as @pratikjagrut is also looking into this.

**Pull request merged**